Automattic/wp-calypso
# Plugins Browser: Redux :allthethings:

Calypso's plugin browser was still using flux for its data. This PR migrates all the data structures used for it to our redux state tree.

A couple of disclaimers:
- Sites are still being read from the old sites-list. The browser uses some selectors that are still not migrated. This will be updated once we migrate everything from sites-list.
- there's some logic in the action that fetches the plugin lists from the API. It's the logic related with controlling when a page can be requested and when we should skip it (for example, because there's already a request for this page going on). I tried to add this logic to the state tree instead, but since we are using the infiniteScroll mixin for this, the way infiniteScroll triggers update events conflicts with the order where the data is updated in the state tree.
  For example, once you request page 1, we dispatch an event to update the state tree value "lastPageFetched", so the next time we try to load the next page, we know that we should call the action for the second page... but it doesn't mind, because by then, infiniteScroll would already have called the 'fetchNextPage' method 10 times, all of the for the first page. Storing this "lastPageCalled" as a variable within the actions module allows us to increment it immediately after the action is called and not depend on the events firing in the correct order.
# How to test
1. Go to http://calypso.localhost:3000/plugins/browse. You should see something like this
   ![image](https://cloud.githubusercontent.com/assets/1554855/19516482/7e109f52-95fc-11e6-8258-f0b954c11e24.png)
2. Browse the different categories. Make sure that next pages are loaded and new plugins are appended when you scroll down
3. Search for any word. Make sure that the results you get make sense and you can scroll down and load further pages.
4. Everything should work the same than it currently does in https://wordpress.com/plugins/browse

ping @enejb @ryelle @tyxla
